fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java

```diff
@@ -71,40 +74,143 @@ public class MvRewritePreprocessor {
     private final ConnectContext connectContext;
     private final ColumnRefFactory queryColumnRefFactory;
     private final OptimizerContext context;
-    private final OptExpression logicOperatorTree;
 
     public MvRewritePreprocessor(ConnectContext connectContext,
                                  ColumnRefFactory queryColumnRefFactory,
-                                 OptimizerContext context,
-                                 OptExpression logicOperatorTree) {
+                                 OptimizerContext context) {
         this.connectContext = connectContext;
         this.queryColumnRefFactory = queryColumnRefFactory;
         this.context = context;
-        this.logicOperatorTree = logicOperatorTree;
     }
 
-    public Set<MaterializedView> getRelatedAsyncMVs(Set<Table> queryTables) {
+    public void prepare(OptExpression optExpression) {
+        try (Timer ignored = Tracers.watchScope("preprocessMvs")) {
+            Set<Table> queryTables = MvUtils.getAllTables(optExpression).stream().collect(Collectors.toSet());
+            logMVPrepare(connectContext, "Query input tables: {}", queryTables);
+
+            try {
+                Set<MaterializedView> relatedMVs =
+                        getRelatedMVs(connectContext, queryTables, context.getOptimizerConfig().isRuleBased());
+                Set<Pair<MaterializedView, MvPlanContext>> validMVs = filterValidMVs(connectContext, relatedMVs);
+                prepareRelatedMVs(queryTables, validMVs);
+            } catch (Exception e) {
+                // MV's prepare should not affect query's process, this may be caused by MV's concurrent process.
+                List<String> tableNames = queryTables.stream().map(Table::getName).collect(Collectors.toList());
+                LOG.warn("Prepare query tables {} for mv failed: {}", tableNames, e);
```

Comment:
Why not throw exception, what if exception happens

Reply:
throw exception just like before.